src/new_api_v04/predicate.rs

```diff
@@ -0,0 +1,57 @@
+//! Predicates to define constraints and patterns.
+//!
+//! A predicate is a boolean-valued function, to be evaluated on some input
+//! data and variable bindings.
+//!
+//! In theory, a binary predicate is thus a function D x U x U -> Bool.
+//! However, to simplify using predicates in practice, we distinguish two types
+//! of predicates:
+//!
+//! 1. [FilterPredicates] The "vanilla" case. Given data and variable bindings,
+//!    evaluate the predicate and return a boolean value.
+//! 2. [AssignPredicates] The "auto-complete" case. Given a binding for <var1>
+//!    and input data, the predicate returns a set of values for <var2> that
+//!    would satisfy the predicate.
+//!
+//! Note that formally, an `AssignPredicate` is just a `FilterPredicate` with a
+//! more efficient way of resolving the constraint: it can always be viewed as a
+//! `FilterPredicate` by calling `assign_check` and then checking that the binding
+//! for <var2> is in the returned set.
+
+use std::collections::HashSet;
+
+/// A predicate for pattern matching.
+pub enum Predicate<AP, FP> {
+    Assign(AP),
+    Filter(FP),
+}
+
+/// A binary predicate that can be queried to return valid RHS bindings.
+///
+/// A predicate of the form `<var1> pred <var2>`. Given a binding for <var1>,
+/// the predicate can return a set of valid values for <var2> for the given
+/// input data.
+pub trait AssignPredicate {
```

Comment:
We could do
```rust
pub trait AssignPredicate : FilterPredicate
```
So assign values can always be used to check if an instantiation is valid too.

We can always do this by checking each element of `check_assign` (as long as `U: PartialEq`), but I imagine it can always be done more efficiently.